### PR TITLE
Add self attentive sentence embedding

### PR DESCRIPTION
**Pull request by kenjewu**
## Description ##
 use [GluonNLP](https://gluon-nlp.mxnet.io/index.html) to reproduce the model structure in [A Structured Self-attentive Sentence Embedding](https://arxiv.org/pdf/1703.03130.pdf) and apply it to [Yelp Data's review star rating data set](https://www.yelp.com/dataset/challenge) for classification.

## Implemented
1. Attention mechanism proposed in the original paper.
2. Punishment constraints to ensure diversity of attention.
3. Parameter pruning proposed in the appendix of the paper.
4. Gradient clip and learning rate decay.
5. SoftmaxCrossEntropy with category weights

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
Add example: Implementation of Structured-Self-Attentive-Sentence-Embedding

### CODE REVIEW

**Comment by szha** [comment]
@sxjscience seems that the PR is still running into the issue of nmt example not finding modules. Could you take a look?

**Comment by mli** [comment]
Job PR-343/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-343/4/index.html

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+model.initialize(init=init.Xavier(), ctx=ctx)
+model.hybridize()
+
+# Attach a pre-trained glove word vector to the embedding layer
+model.embedding_layer.weight.set_data(vocab.embedding.idx_to_vec)
+# fixed the layer
+model.embedding_layer.collect_params().setattr('grad_req', 'null')
+```
+
+Using r attention can improve the representation of sentences with different semantics, but if the value of each line in the attention matrix A (r-byn) is very close, that is, there is no difference between several attentions. In the subsequent M = AH, the resulting M will contain a lot of redundant information.
```
Comment (bikestra) [inline docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: >In the subsequent M, = AH ->  Subsequently, in M = AH,

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+        batch_x = batch_x.as_in_context(ctx)
+        batch_y = batch_y.as_in_context(ctx)
+        batch_pred, att = model(batch_x)
+        if loss_name == 'sce':
+            l = loss(batch_pred, batch_y)
+        elif loss_name == 'wsce':
+            l = loss(batch_pred, batch_y, class_weight, class_weight.shape[0])
+
+        # penalty
+        temp = nd.batch_dot(att, nd.transpose(att, axes=(0, 2, 1))
```
Comment (bikestra) [inline docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: `temp` is inappropriate since it delivers no information. Why not something like `diversity_penalty`?

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+
+    print('Train on ', ctx)
+    for epoch in range(1, num_epochs + 1):
+        start = time.time()
+        train_loss = 0.
+        total_pred = []
+        total_true = []
+        n_batch = 0
+
+        for batch_x, batch_y in data_iter_train:
```
Comment (bikestra) [inline docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: There is quite a bit of copy & paste between `evaluate()` and `train()`, which should be consolidated. Note that you can use something like 
```
            with ExitStack() as stack:
                if is_train:
                    stack.enter_context(autograd.record())
```
to make sure you enter `autograd.record()` context only in training.

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+clip = 0.5
+nepochs = 4
+
+trainer = gluon.Trainer(model.collect_params(), optim, {'learning_rate': lr})
+
+if loss_name == 'sce':
+    loss = gluon.loss.SoftmaxCrossEntropyLoss()
+elif loss_name == 'wsce':
+    loss = WeightedSoftmaxCE()
+    class_weight = nd.array([3.0, 5.3, 4.0, 2.0, 1.0], ctx=ctx)
```
Comment (bikestra) [inline docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: Where do these numbers come from?

**Comment by mli** [comment]
Job PR-343/6 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-343/6/index.html

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+train_dataset, train_data_lengths = preprocess_dataset(train_dataset)
+valid_dataset, valid_data_lengths = preprocess_dataset(valid_dataset)
+```
+
+```{.python .input}
+# create vocab
+train_seqs = [sample[0] for sample in train_dataset]
+counter = nlp.data.count_tokens(list(itertools.chain.from_iterable(train_seqs)))
+
+vocab = nlp.Vocab(counter, max_size=10000, padding_token=None, bos_token=None, eos_token=None)
```
[the remark this answers is not in the document]
Comment (kenjewu) [reply docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: Yes, the standard way of writing is what you said.  because my data in this experiment is filled with 0, 0 corresponds to unk in the vocabulary, so padding_token is set to None.

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+
+    def hybrid_forward(self, F, x):
+        # x shape: [batch_size, seq_len, embedding_width]
+        # ut shape: [batch_size, seq_len, att_unit]
+        ut = self.ut_dense(x)
+        # et shape: [batch_size, seq_len, att_hops]
+        et = self.et_dense(ut)
+
+        # att shape: [batch_size,  att_hops, seq_len]
+        att = F.softmax(F.transpose(et, axes=(0, 2, 1)), axis=-1)
```
[the remark this answers is not in the document]
Comment (kenjewu) [reply docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: Yes, ''-1'' represents the last axis.

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md (hunk 4 above)
Comment (kenjewu) [reply docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: The value of class_weight is obtained by counting data in advance. It can be seen as a hyperparameter. Different data should have different class_weight.

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md (hunk 3 above)
Comment (kenjewu) [reply docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: Because there are gradient calculations, cropping, parameter updates, learning rate reduction, and output progress in training. But not in the evaluation. I followed the train and evaluation in the gluonbook and made a simple modification.

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+        batchify_fn=batchify_fn)
+    return train_dataloader, valid_dataloader
+
+train_dataloader, valid_dataloader = get_dataloader()
+```
+
+## Model Structure
+
+In the original paper, the representation of the sentence is: firstly, the sentence is disassembled into a list corresponding to the word, then the words are unrolled in order, and the word vector of each word is calculated as the input of each step of the [bidirectional LSTM neural network layer](https://www.bioinf.jku.at/publications/older/2604.pdf)[3]. Taking the output of each step of the bidirectional LSTM network layer, a matrix H is obtained. Suppose the hidden_dim of the bidirectional LSTM is `u`, the word length of the sentence is `n`, then the dimension of the last H is `n-by-2u`.  For example, the sentence "this movie is amazing" would be represented as:
+![](../images/Bi-LSTM-Rep.png)
```
Comment (szha) [inline docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: this should simply be ``

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+
+Attention is like when we are looking at things, we always give different importance to things in the scope of the perspective. For example, when we are communicating with people, our eyes will always pay more attention to the face of the communicator, not to the edge of other perspectives. So when we want to express the sentence, we can pay different attention to the output H of the bidirectional LSTM layer.
+![](../images/attention-nlp.png)
```
Comment (szha) [inline docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: likewise

File: docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md (hunk 3 above)
Comment (bikestra) [reply docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: yes, there are additional operations, but you can make it optional by making `is_train` a parameter. I understand that other notebooks don't do a good job in code sharing between training and evaluation, but that shouldn't prevent us from writing better code; we should set up a better standard.

Comment (kenjewu) [reply docs/examples/self_attentive_sentence_embedding/self_attentive_sentence_embedding.md]: You are right, I have modified this problem to minimize the writing of duplicate code, but I don't understand ExitStack very well.

**Comment by mli** [comment]
Job PR-343/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-343/11/index.html

**Comment by szha** [comment]
I force-pushed the wrong branch to the upstream. @kenjewu is creating another PR with the same updated content.
